ppa/archive/management/commands/gensim_serialize.py

```diff
@@ -0,0 +1,292 @@
+        :param doc_id_field: Solr field name for document identifier. Default 'source_id'.
+        :param doc_content_field: Solr field name for document content. Default 'content'.
+        :param page_order_field: Solr field name for page ordering. Default 'order'.
+        :param doc_limit: Max no. of documents to process. The default of -1 means we process ALL documents found.
+        :param doc_ids: List of document identifiers to process. If None, ALL documents found are processed.
+        :param preprocess_fns: A list of single-argument functions to use as preprocessors.
+            See the module gensim.parsing.preprocessing for some typical preprocessing functions.
+        :param pbar: A boolean indicating whether to display a progress bar during corpus generation.
+        """
+        self.name = name
+        self.client = client
+        self.collection = collection
+        self.doc_id_field = doc_id_field
+        self.doc_content_field = doc_content_field
+        self.page_order_field = page_order_field
+        self.doc_limit = doc_limit
+        self.preprocess_fns = preprocess_fns or []
+
+        self.dictionary = Dictionary()
+
+        # doc_id -> dict of k->v mappings
+        # NOTE: We cannot use 'metadata' as GenSim mangles this attribute internally!
+        self._metadata = {}
+
+        # list of strings, populated on first doc retrieval
+        self.metadata_field_names = None
+
+        results = self.client.query(
+            collection=self.collection,
+            query={
+                'q': '*:*',
+                'facet': 'on',
+                'rows': 0,
+                'facet.field': doc_id_field,
+                'facet.limit': self.doc_limit
+            }
+        )
+
+        """
+        An OrderedDict of doc_id => page count mapping
+        An OrderedDict is important here in case we want to save document-level metadata, in which case rows of
+        metadata would be in the same order as the BoW-vectors returned by this object's iterator.
+        """
+        self.page_counts = results.get_facets()['source_id']
+        self.doc_ids = doc_ids or self.page_counts.keys()
+        self.doc_count = len(self.doc_ids)
+        self.pbar = ProgressBar(redirect_stderr=True, max_value=self.doc_count) if pbar else NullBar()
+
+    def __iter__(self):
+
+        for doc_id in self.doc_ids:
+
+            if doc_id not in self.page_counts:
+                logger.warning('Unknown page count for doc {}. Skipping.'.format(doc_id))
+                continue
+
+            result = self.client.query(
+                collection=self.collection,
+                query={
+                    'q': '{}:{}'.format(self.doc_id_field, doc_id),
+                    'order': '{} asc'.format(self.page_order_field),
+                    'rows': self.page_counts[doc_id]
+                }
+            )
+
+            metadata_docs = [d for d in result.docs if d['item_type'] == 'work']
```

Comment:
So I did some experiments on this, and it turns out that querying Solr just on a document id is _faster_ than querying it on the document id with `item_type:page` (not even counting the time it would take to run **another** query in the latter case for `item_type:work`, to get the associated metadata). This makes sense to me, since querying on a single index is faster than querying on two.

Since the runtime for each document id is dominated by Solr, any filtering we do on the Python end is turning out to have negligible impact.